src/natural_history_manager.rs

```diff
@@ -0,0 +1,572 @@
+/// when there are at least two samples above and below `xp`. Otherwise, it uses linear extrapolation at the tails.
+/// Assumes that function samples are sorted so that the `xs` are in ascending order.
+fn interpolate(xs: &[f64], ys: &[f64], xp: f64) -> f64 {
+    let upper_window_index_option = xs.iter().position(|&x| x > xp);
+    // We need to check whether a point was found. If it wasn't, it means that all values
+    // in `xs` are less than `xp`. We have to use an alternative extrapolation strategy.
+    let upper_window_index = match upper_window_index_option {
+        None => {
+            // We are above the range of the `xs` samples, so we must extrapolate. We default to
+            // linear extrapolation using the last two values in `xs`.
+            let traj_len = xs.len();
+            return linear_interpolation(
+                xs[traj_len - 2],
+                xs[traj_len - 1],
+                ys[traj_len - 2],
+                ys[traj_len - 1],
+                xp,
+            );
+        }
+        Some(i) => match i {
+            // We only have one point in `xs` below `xp`, so we must also default to linear extrapolation.
+            1 => return linear_interpolation(xs[0], xs[1], ys[0], ys[1], xp),
+            // Else, we use cubic spline interpolation.
+            // The index can never return 0 because we handle that case at the beginning of the function.
+            _ => i,
+        },
+    };
+    // If our interpolation point is in the range of the `xs` samples, we can use cubic spline interpolation.
+    cubic_spline_interpolation(
+        xs[(upper_window_index - 2)..=(upper_window_index + 1)]
+            .try_into()
+            .unwrap(),
+        // Since `upper_window_index` is the third index in the trajectory, we need to subtract 2 to get the first index.
+        ys[(upper_window_index - 2)..=(upper_window_index + 1)]
+            .try_into()
+            .unwrap(),
+        xp,
+    )
+}
+
+/// Fits a line between two points and returns the value of the line at a given x-value `xp`.
+/// Sensibly handles equal x-values by returning `y1` for `xp < x1`, `y2` for `xp > x2`, and
+/// `(y1 + y2) / 2` for `xp = x1 = x2`.
+fn linear_interpolation(x1: f64, x2: f64, y1: f64, y2: f64, xp: f64) -> f64 {
+    // At the tails of the CDF, the CDF moves very slowly, so with numerical imprecision, x2 may equal x1.
+    #[allow(clippy::float_cmp)]
+    if x2 == x1 {
+        // The only sensible thing to return is the bounds.
+        if xp < x1 {
+            return y1;
+        } else if xp > x2 {
+            return y2;
+        }
+        return (y1 + y2) / 2.0;
+    }
+    (y2 - y1) / (x2 - x1) * (xp - x1) + y1
+}
+
+/// Fits a cubic spline between four points and returns the value of the spline at a given x-value `xp`.
+/// Requires that `xp` be between the second and third ordered points in the `xs` array.
+fn cubic_spline_interpolation(xs: &[f64; 4], ys: &[f64; 4], xp: f64) -> f64 {
```

Comment:
I'm not a math guy but given the expected density of these points, is cubic really that much better than linear? 

My instinct would actually be to do all this interpolation once at loading time down to some precision limit (.1 days?) and then just look up the closest value and take it.

Reply:
Thanks. I initially used just linear interpolation and I noticed that I needed a much higher density of points than I anticipated, prompting the cubic interpolation. Basically:

1. Since CDFs can be either concave up or concave down at a given point, linear interpolation will always overestimate or underestimate the CDF, respectively. This results in consistent bias in the estimate of the inverse CDF, though the direction of that bias may change. By fitting to more points, cubic interpolation largely fixes this problem.
2. When doing linear interpolation, I had to provide CDF values down to 0.02 days in order to reproduce an exponential distribution with a KS stat of less than 0.01. That's a lot of points, and the model took a long time to start up. Comparatively, when using cubic interpolation, I could easily get away with CDF values down to 0.2 days, resulting in nearly a 5x speed up in just running the tests.
3. I don't think we should expect users to provide a given density of points. I agree that interpolation comes with a certain degree of "proceed with caution" but I also think the degree/type of interpolation used here is rather straightforward and standard.

I need to think more about doing the interpolation once -- I like the implications of this idea for the structure of the model, but my initial intuition is that you would end up not needing the interpolated values over a large range of the CDF in the actual simulation.

Reply:
I'll defer to @jasonasher about the math parts of it, but a few thoughts on the software engineering.

I don't think we should decide what to do here based on the test running time:

* Your current trajectory file is 4K lines. Reading 40K lines worth of CSV ought to be really fast, so if it's not, then I worry we have a different problem.
* If for some reason it's not fast, can we load it once into a fixture and cache it?

It seems like we're going to need to be doing interpolation a lot (for essentially each infection attempt, right?) so there is a lot of incentive to make that as fast as possible, even at considerable upfront cost, especially if that cost can be borne outside the simulation.